### ISSUE TITLE
Getting "Fachmann" Data updates again

### ISSUE THREAD

**Issue by tomohnejerry** [iobroker-community-adapters/ioBroker.wolf-smartset#242]
Hello there,
first I'll like to thank you for this great adapter. 
Since a few months I noticed that the "fachmann" values are only updated if I login on the wolf-smartset webinterface and go into the fachmann corner.
Is there a possibility to get this values updatet as usual? I'm pretty shure that there was a change on wolf, because I used the stable v1.0.0 . I also tried the 1.1.0 but it seems it had no impact.
Thanks alot

**Comment by Apollon77** [comment]
We jad this discussion last time in an other thread ... As I tested, e,g, the data was updated ... so in fact the adapter qeries the same values as the website in 1.x ... so no idea whats the difference. Becaus I do  not have such a device I can not say more

**Comment by tomohnejerry** [comment]
As i wrote the data from the unprotected entries gets pulled well. But only the Fachmann data which is protected by a 4 digit key gets only updated when I log into this area in the web ui. 
If I can help you with some data please feel free to ask
Thansk

**Comment by tomohnejerry** [comment]
Seems like Wolf changed the way they send data to the Smartset cloud. Because the Fachmann-Data gets only updated when I login to this specific area. As the data gets fetched from the bm-2 (heating system) online (takes some seconds) I see the update of the values live in the web interface wolf and so the data in my ioBroker instance does.  
I hope this helps to get a solution if possible

**Comment by Savari-dev** [comment]
I like the Wolf-Smartset-Adapter as well - thank you for the implementation :-). I am using it for receiving efficency information for my heatpump and displaying them in iobroker.
Unfortunately i am suffering from this issue as well. Obviously Wolf has mixed Push and Pull techniques to update the smartset database from the installations. The Fachmannebene needs a login and afterwards triggers the update.
Unfortunately the "Fachmanebene"-Data includes statistic data which are relevant to optimize the effiviency of the heatings. I assume that only the Wolf-Development Team can solve this problem. Maybe it could be a simple solution to shift some data from "Fachmannebene" to "Benutzerebene" in order to have regular updates. In my case i am only interested in powerconsumption data.
Is there an actual state for this issue? 
thank you in advance,

**Comment by SpatialJ** [comment]
any  updates on this issue? Was anyone able to find a solution or workaround?

**Comment by Savari-dev** [comment]
@SpatialJ: I had not used this Adapter for a while but i retried yesterday. I have received all statistics-data that i needed from wolf-smartset portal. For me the problem seems to be fixed.

**Comment by SpatialJ** [comment]
@Savari-dev  I am using it as well, and the issue is still here :-/

**Comment by mcm1957** [comment]
> @Savari-dev I am using it as well, and the issue is still here :-/

Please add version informtaion you are using:
adapter
js-controller
node
os

Please attach a log with LEVEL set to DEBUG

**Comment by SpatialJ** [comment]
> > @Savari-dev I am using it as well, and the issue is still here :-/
> 
> Please add version informtaion you are using: adapter js-controller node os
> 
> Please attach a log with LEVEL set to DEBUG

Hi,
this is from the iobroker info:
`Platformoperating system:linux
Architecture:x64
CPUs:2
Speed:0 MHz
Model:Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
RAM:3.8 GB
System uptime:25 d. 04:55:22
Node.js:v20.18.0
time:1734341784142
timeOffset:-60
NPM:10.8.2
adapters count:549
Disk size:29.5 GB
Disk free:14.2 GB
Active instances:10
Path:/opt/iobroker/
Uptime:25 d. 04:53:44
alive:true
_nodeCurrent:20.18.0
_nodeNewest:20.18.1
_nodeNewestNext:20.18.1
_npmCurrent:10.8.2
_npmNewest:10.8.2
_npmNewestNext:10.8.2`

The Wolf Smartset Adapter Version is:
`Wolf SmartSet

Installed instances:
1
Available version:
1.2.3
Installed version:
1.2.3`


The Logs look similar for each day, for example:
`
2024-12-14 03:45:46.326 - warn: wolf-smartset.0 (301) API REFRESH TOKEN ERROR: login_required: undefined
--
2024-12-14 03:45:46.338 - warn: wolf-smartset.0 (301) Error: login_required: undefined
at WolfSmartSet._refresh (/opt/iobroker/node_modules/iobroker.wolf-smartset/lib/wss.js:455:11)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-12-14 03:45:47.192 - info: wolf-smartset.0 (301) Initialization with Login and new Session done!
2024-12-14 04:35:17.982 - warn: wolf-smartset.0 (301) State value to set for "wolf-smartset.0.Fachmann.Heizgerät.Parameter_HG.419_Einstellungen.27010400001" has value "2" less than min "5"
2024-12-14 05:42:48.947 - info: admin.0 (160) Adapter rating updated
2024-12-14 07:05:44.448 - info: admin.0 (160) Request actual repository...
2024-12-14 07:05:44.629 - info: host.iobroker Updating repository "stable" under "http://download.iobroker.net/sources-dist.json"
2024-12-14 07:05:47.106 - info: admin.0 (160) Repository received successfully.
2024-12-14 13:45:48.990 - warn: wolf-smartset.0 (301) API REFRESH TOKEN ERROR: login_required: undefined
2024-12-14 13:45:48.990 - warn: wolf-smartset.0 (301) Error: login_required: undefined
at WolfSmartSet._refresh (/opt/iobroker/node_modules/iobroker.wolf-smartset/lib/wss.js:455:11)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-12-14 13:45:49.858 - info: wolf-smartset.0 (301) Initialization with Login and new Session done!
2024-12-14 23:45:51.608 - warn: wolf-smartset.0 (301) API REFRESH TOKEN ERROR: login_required: undefined
2024-12-14 23:45:51.608 - warn: wolf-smartset.0 (301) Error: login_required: undefined
at WolfSmartSet._refresh (/opt/iobroker/node_modules/iobroker.wolf-smartset/lib/wss.js:455:11)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-12-14 23:45:52.480 - info: wolf-smartset.0 (301) Initialization with Login and new Session done!
`

**Comment by Savari-dev** [comment]
@mcm1957 i just have changed the log-level to debug. I will give you the resulting file tomorrow.
Here are my versions:
Wolf-Smartset Adapter: 1.2.3
js-controller: 6.0.11
node.js: 20.17.0
os linux bookworm
Please note that i am using two parameters out of the Fachmann-Menue:
wolf-smartset.0.Fachmann.Heizgerät.Heizgerät.413_Statistik.27016400001 Verbrauch aktuelles Jahr
wolf-smartset.0.Fachmann.Heizgerät.Heizgerät.413_Statistik.27016500001 Waermemenge aktuelles Jahr
Both of these are updated by Wolf-Smartset-Portal irregularly. It seems that one time in the early morning the update is reliable.
I am using this data for daily , monthly and yearly statistics, so the update once a day is sufficient for me.

**Comment by Savari-dev** [comment]
Here is my logfile from today
[iobroker.2024-12-17.log](https://github.com/user-attachments/files/18162934/iobroker.2024-12-17.log)

**Comment by flingo64** [comment]
> Here is my logfile from today [iobroker.2024-12-17.log](https://github.com/user-attachments/files/18162934/iobroker.2024-12-17.log)

The log does not contain der `API GET GUI DESC.:` line we are looking for. Also, the log looks like copied from screen, not downloaded as log. Can you please proceed as follows:
- in the Wolf adapter instance settings set Log level to Debug, do not(!!!) check 'No reload'
![grafik](https://github.com/user-attachments/assets/9cb83cc2-bb10-4e1c-81ce-d8ae3b09eb61)
- Close adapter instance, wait until adapter has reloaded
- go to Logs and download logs for the current date
![grafik](https://github.com/user-attachments/assets/20cbc378-a363-4ca8-881b-30efec30b91b)
- upload the downloaded log as whole or extract the  `API GET GUI DESC.:` line, which ends at the next timestamp (will look like be several lines in your editor)
- in the Wolf adapter instance settings set Log level to Info agian , do not(!!!) check 'No reload'

Thnx

**Comment by flingo64** [comment]
> I have found an appropriate line in the debug log from yesterday. [API GET GUI DESC Log.txt](https://github.com/user-attachments/files/18168579/API.GET.GUI.DESC.Log.txt)

Excellent! Here are the inconsistencies I found (looks pretty much the same as my CHA-07, see [issue #281](https://github.com/iobroker-community-adapters/ioBroker.wolf-smartset/issues/281#issuecomment-2546631979):

---
| ValueId | Name | Group | ValueState | Value | MinValueCondition | MaxValueCondition | MinValue | MaxValue | IsReadOnly | IsExpertProtectable | 
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| 27010400001 | Hysterese Kühlbetrieb | 419_Einstellungen | 1 | 2 | 0.5 | 10 | 5 | 10 | false | false | 
| 27012800001 | Aktueller Energiepreis ZWE | 419_Einstellungen | 1 | 6 | 0.1 | 99.9 | 1 | 999 | false | false | 
| 27012900001 | Aktueller Strompreis | 419_Einstellungen | 1 | 21 | 0.1 | 99.9 | 1 | 999 | false | false | 
| 22003000000 | PI-Regler Raumtemperatur Tn | 400_Anlagenparameter | 1 | 1 | 0.1 | 20 | 1 | 20 | false | false | 

I'm currently setting up a ioBroker-dev so I can test my fix. I'll let you know when it's available for testing...

**Comment by mcm1957** [comment]
Please let us know if the problem still exists with current 1.2.4

**Comment by SpatialJ** [comment]
How can i get 1.2.4?
In my installation it says:
![image](https://github.com/user-attachments/assets/6a2fda52-5922-4aac-8046-56243e7491fa)

**Comment by mcm1957** [comment]
1.2.4 is available at LATEST / BETA Repository since app 2 weeks and will be available at STABLE repository tomorrow.

**Comment by SpatialJ** [comment]
I tried with 1.2.4 today, but the problem remains. The values from the "Fachmann-Ebene" are only retrieved when i open it in a separate browser. It then keeps receiving new values for about 10 minutes, then it stops until i reopen the page in the browser. You can see that in the attached logs. No Fachmann-values were received between 12:00 and 13:00. After i logged back in at 13:02 they are received again.
[iobroker.2025-01-08.html.zip](https://github.com/user-attachments/files/18347100/iobroker.2025-01-08.html.zip)

**Comment by flingo64** [comment]
Please **remove all old objects** under:
-     238_Warmwasser
-     410_Einstellungen_und_Anzeigen
-     400_Anlagenparameter
-     419_Einstellungen
-     481_ISM
 They were replaced by objects in new subfolders as described in this [post](https://github.com/iobroker-community-adapters/ioBroker.wolf-smartset/issues/281#issuecomment-2552365244) und so, their object definition won't be updated anymore.

After removing the old objects restart the adapter.

**Comment by SpatialJ** [comment]
Thanks for your suggestions! I did all that but the the behavior remains the same. i only get fachmann-values when i have a browser session open at the same time. After i close the browser it will keep receiving values for 10 minutes then it stops until i log back in. This morning for example, after making the suggested changes, i closed the browser at 8:41 and got updates until 8:51 then it stopped. After loggin back in at around 9:00 it started receiving values again.
I attached the logfile (there was a problem with smartset portal until this morning)
[iobroker.2025-01-09.txt](https://github.com/user-attachments/files/18358264/iobroker.2025-01-09.txt)

**Comment by flingo64** [comment]
After reading the whole thread, I'm sure the fix in v1.2.4 is not related to this issue. 

Anyway, I was able to view your logfile (after renaming it to .html), but I wasn' t able to identify which particular datapoints you are missing between 8:51 and 9:00. Could you give me an example value id. I would then check how often I get it in my configuration. 

Interesting, though: I'm doing a quite elaborated monitoring in Grafana (using the InfluxDB adapter of ioBroker) of my CHA-07 and I'm getting all value of interest with a reasonable frequency. see below.
![image](https://github.com/user-attachments/assets/1a2f9556-b967-4ac2-9b28-cbb6d3285c86)
![image](https://github.com/user-attachments/assets/c77aaab8-fc9e-4f81-b291-004d4d64c7df)

**Comment by SpatialJ** [comment]
Sure,
in particular im looking for
27004100001
27004200001
27006300001
27006400001
27006500001

funny enough im using it for the same reason, to monitor my cha7 with influxdb/grafana.
You can see the gap/missing values in my Kältekreispanel as well. I just logged back into the fachmann menü on the smartset portal, so the values start to appear agagin in the panel :-)
![image](https://github.com/user-attachments/assets/c6027bf7-800c-4647-b2e1-d13458b2c501)

**Comment by Knuppes** [comment]
I got the same problem. 

From my point of view, it seems that the wolf-portal allows access to the source ip: My iobroker is running on a raspi, Fachmann-Menu is opened on my Mac. ... and it works as described above.

So, as a workaround, it would be enough to open the Fachmann-Menu every 9 minutes with curl in a loop from any pc in the home-network... I will give this a try if I can find an hour at the weekend.

Another thing: @SpatialJ @flingo64 : Would you both mind to share your Grafana Dashboards somehow? I'm using influxdb and Grafana as well, but your stuff seems much more impressive...

**Comment by flingo64** [comment]
I think I got it.

After demystifying/decoding the API URLs used by the adapter and some Browser sniffing on the Wolf Smartset Portal I found the API to authorize as Expert (Fachmann). So, now when authenticating/re-authenticating against the Wolf server the adapter will also do the Expert Login within that session. Unfortunately, Wolf still did not return the above mentioned Fachmann values. The reason for that I found today: when issueing the first GetParameterValues call you have to tell Wolf that this is the very first request (LasstAccess = null) and that you just changed the GuiId (GuidChanged=true) and then: Bingo!

The adapter is now up for half an hour and still is getting frequent updates of the above mentioned values (no Wolf Smartset Portal session in parallel)!
![grafik](https://github.com/user-attachments/assets/8296e3df-30f0-4e40-b14b-e7a6778cf993)
I'll watch it for a while on my dev-server and install it on my production system tomorrow. If it all work well, I'll push der changes to my Github fork, so anyone can test it upfront before creating the pull request to this repo.

Stay tuned ...

**Comment by SpatialJ** [comment]
Nice! Thanks for your effort! Im looking forward to 1.2.5 :-)

**Comment by flingo64** [comment]
Unfortunately, today all looks like before. :-( Although the Expert Login works fine. I'm not getting the above mentioned values refreshed after the initial full pull. I tried a thousand of small changes but to no avail. 
Even worse, my IP is currently obviously blocked during authentication already, so I can't enter der Web Portal or the app via fixed line. Using mobile network. I can login with the app. 
So, I have to reboot the router to get a new IP, but my wife will kill me, if I'm doing it now...

**Comment by flingo64** [comment]
![grafik](https://github.com/user-attachments/assets/065c1b0e-9c58-4d9c-b1b4-90157844ea04)

:-)

Still some work to do, but that looks promising!

**Comment by flingo64** [comment]
by issueing the expert login and and doing expert re-login and some more re-initializations every hour I got the expert values for about 20 hours. After that all over sudden, I got:
`API GET GUI DESC. ERROR: Request failed with status code 400`
and after that no more refreshing of expert value. Even a complete restart of the adapter did not solve the issue. In fact, even the Wolf Smartset Portal did no longer show values in the Expert section. A few hours later, also the standard values returned w/ no value and state = 0 (no change). So, I switched back to Non-Expert Mode to receive at least the standard values, 12 later I tried again the Expert mode and now I get them again ...

I suspect that besides the proven 60 sec minimum interval limitation:
`API GET USERINFO: {"IsPasswordReset":false,"UserName":"xxx","UserSalutationType":2,"Firstname":"xxx","Surname":"xxx","CultureInfoCode":"de-DE","TwoLetterCountryCode":"DE","Email":"xxx@yyy","MaxSystemBusSamplingRateSec":60}`
there might be a limit on total object values delivered per day or similar. To handle such a limitation it requires a more sophisticated handling of which values (or bundles) have to be queried how often. This may take a while ...

![Image](https://github.com/user-attachments/assets/818ddadf-5b51-455b-ac45-afa571e9983f)

**Comment by flingo64** [comment]
I got stable delivery of expert values for now almost two days now. What I changed so far:
- the list of requested values now excludes roughly 100 settings values which are changed very rarely
- the poll interval was changed from 60s to 120s (weniger ist mehr). 
The ongoing value delivery from the Wolf server may proof the suspected limitations of values per day or ssimilar.

To get a insight into the server behaviour I added some kind of API profiling, so one can see how many values were requested from the adapter and how many were returned from Wolf server. 

In the figure below, the API profiling is shown in the last diagram:
- starting in standard mode, the adapter requests all values in Bundle 1000 (116 values)
- shortly after that I switched to expert mode, so the adapter asks for all values in Bundle 1000 and 3100 (168 values)
- every 2 hours the adapter does re-login for expert mode and reinitialzies the query, so Wolf will send all requested values
- except for the initialization, Wolf does not return all requested params, but probably only those with changed values
- I changed the poll interval yesterday at 16:00 from 60s to 120s: the number of returned values is now more constant
![Image](https://github.com/user-attachments/assets/155596b7-d160-4cfc-8088-79d35de1e873)

Lessions learnt: the more values you ask for, the longer the poll interval has to be

Regarding bundles: Wolf assigns each value to a specific bundle and the Portal specifies the according BundleId in its requests to the server. I'm still not sure if the BundleId is really important in the poll request, because the list of requested values are also given in the poll request. Anyway, the next release will show the BundleIds of all values the name of its parent channel. 

![Image](https://github.com/user-attachments/assets/018a7f59-497e-459f-8a58-9db3b3b98422)

I will do another test with the adapter requesting all values (not just Bundle 3100 and 1000) with a 2 minute interval and see how that works out.

**Comment by SpatialJ** [comment]
Thanks for remaining  persistent with this issue :-)
What a werid API though :-)

**Comment by flingo64** [comment]
Requesting all values works well with a 2 minute poll interval. In fact, I switched back to Non-Expert mode yesterday, so now the 2-hourly Expert re-login is not performed anymore, only the openid re-auth w/ username/password every 10 hours is performed as before. Curiously enough, I'm still getting expert values (die Geister, die ich rief...)!  So in fact, I'm almost back to the first solution I had already 2 weeks ago, except the poll interval.

Another issue irritates me, because I'm not sure whether it is related to the adapter modifications or a Wolf server issue: the following values are returned periodically, but have **not been updated since 2025-01-21 04:21**:
- wolf-smartset.0.Benutzer.Heizung.212_Statistik_Wärmeerzeuger 1.27017500001 (Verbrauch Vortag)
- wolf-smartset.0.Benutzer.Heizung.212_Statistik_Wärmeerzeuger 1.27017600001 (Energiemenge Vortag)
- wolf-smartset.0.Benutzer.Heizung.212_Statistik_Wärmeerzeuger 1.27017600001 (TAZ Vortag)

Those values can also be found in the Portal and in the App, and they are not changed their either.

Can anybody confirm this issue? 

![Image](https://github.com/user-attachments/assets/213072eb-60a1-464e-9bc0-7549d187ecf8)

**Comment by Knuppes** [comment]
Hi Flingo64,

from my side it looks OK, there are values for the entire last week, despite the fact that I had turned down the heater for the weekend and changed to good old wood-oven.

I attached a file with the "Verbrauch Vortag" for the last week, this other to look also good.

Cheers

Knuppes


[Verbrauch Vortag.csv](https://github.com/user-attachments/files/18554620/Verbrauch.Vortag.csv)

**Comment by flingo64** [comment]
@Knuppes,
thanks for your valueable feedback! So, it seems this issue relates to fact that the adapter is running in expert mode since a week.

New fact on my side: 
- for some reason the Wolf server decided tonight at 3:00 am to stop sending expert values (remember, I turned it off already 2025-01-25 around 20:00). 
- today I got fresh values for the Vortag statistics 
- funny enough: the server restarts sending expert values after the periodic 10-hour Token-refresh, I'm 100% sure the adapter did not do a Expert Login (it would be shown in the adapter log)

This server is driving me nuts! It seems that **the Wolf server won't send Vortag statistics as long as in expert mode**.
The good question though: how can I stop the expert mode?! I'll try to send a API Logout on adapter reload and see if it helps ...

![Image](https://github.com/user-attachments/assets/7faaa931-6e88-4205-b0ac-676d1bcd6c6e)

**Comment by Knuppes** [comment]
@flingo64 

do you have an ordinary DSL(or similar)-connection? I'm thinking about the "Zwangstrennung" which is usually in the middle of the night.

In fact, I wonder that you can get values over night without reconnecting if your external IP changes. I experienced, that a login to Fachmannmenü on my Workstation, enables accessing values of Fachmannmenü on my Raspi. From the Wolf Portal's view, it it the same source IP. ... which is forced to changed every night...

It seems to me, that the portal is doing a strange mixture of OpenID **AND** source IP... and I'm not sure if this is a bug or a feature...

**Comment by flingo64** [comment]
> [@flingo64](https://github.com/flingo64)
> 
> do you have an ordinary DSL(or similar)-connection? I'm thinking about the "Zwangstrennung" which is usually in the middle of the night.
> 

Thanks for the hint: I checked in FritzBox, but the session/IPv4/IPv6 is up and running since 2025-01-12. (btw: Telekom is no longer doing any Zwangstrennungs on (V)DSL since years).

But you are right, it seems the IP address of the client is important in their session management.

**Comment by flingo64** [comment]
It took a quite long time to understand and fix the issue of the server stop sending expert values at some point in time. This is now fixed by the above mentioned pull request. It required heavy code-reworks and after all it includes besides the actual fix a lot of enhancements, to name a few:
- support for level-3 parameters such as time programs, party mode, vacation mode
- detection and marking of obsoleted parameters
- API profiling
- grouping of parameters values
- two independent poll cycle
- detection of changes of the adapter's public IP and Wolf Smartset session re-initializing afterwards 
- completely new Admin UI (instances settings dialog) based on jsonConfig API
- more comprehensive README
The adapter is working fine and stable for weeks. You may check out the [fork](https://github.com/flingo64/ioBroker.wolf-smartset) or wait until the pull request was merged.

Have fun!

**Comment by mcm1957** [comment]
fixed with 3.0.0
thanks @flingo64 
https://github.com/iobroker-community-adapters/ioBroker.wolf-smartset/pull/437

**Comment by SpatialJ** [comment]
@flingo64  thank you for your effort!  Will i be able to just upgrade to the new version or will i need to chnage the configuration?

**Comment by mcm1957** [comment]
As far as I saw coechanges only reentering username / password should be required. But lest wait for feedback from @flingo64

**Comment by flingo64** [comment]
Yes, you will have to re-enter username and password of your Smartset account. After that you will get a list of possible Wolf devices (most likely exactly one). Select and confirm your device selection and the adapter is ready to go. If you want to pull expert values (that's what it all was about) you'll have to enter the Expert Password (the four digits) on the 'Advanced Settings' tab. There's more info on all settings in the README.
